lib/kcapi-kdf.c

```diff
@@ -54,6 +54,20 @@
 #include "kcapi.h"
 #include "internal.h"
 
+#define GCC_VERSION (__GNUC__ * 10000		\
+		     + __GNUC_MINOR__ * 100	\
+		     + __GNUC_PATCHLEVEL__)
+#if GCC_VERSION >= 40400
+# define __HAVE_BUILTIN_BSWAP32__
+#endif
+
+#ifdef __HAVE_BUILTIN_BSWAP32__
+# define be_bswap32(x) (uint32_t)__builtin_bswap32((uint32_t)(x))
```

Comment:
I don't think this is quite right. `__builtin_bswap32()` will always reverse the order of bytes, regardless of byte order, but you only want to do it when `__BYTE_ORDER__ ==  __ORDER_LITTLE_ENDIAN__`. 

I think you will have to nest the `__HAVE_BUILTIN_BSWAP32__` check under the `#elif __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__`. (Yes, it will produce slightly hard-to-read code, but I don't think it can be avoided.)

Reply:
yes you are right